### PR TITLE
[PREVIEW] working on migration to 3.4 api

### PR DESCRIPTION
**Pull request by schernolyas**
HI @DavideD !

I change Eclipse to Idea and prepare first part of changes. 
I just changed DBObject to Document. look my PR please.
Need do:
1) continue migration
2) correct tests
3) check and fix tests

to be continued

### CODE REVIEW

**Comment by schernolyas** [comment]
Hi @DavideD !

I have refactored main classes. 
I see that the PR is better that previus and ... I hope You can make review.

**Comment by schernolyas** [comment]
Hi @DavideD !
Code in the branch is compile only. but ... not works.
Please, review the changes and mark them. I will fix tests and your remarks.

Also ... not close the branch  please

File: mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java [hunk 1]
```diff
@@ -257,46 +265,49 @@ private DBObject getEmbeddingEntity(AssociationKey key, AssociationContext assoc
...
 
-		BasicDBObject projection = getProjection( tupleContext );
+		Document projection = getProjection( tupleContext );
 
-		DBObject query = new BasicDBObject();
-		query.put( ID_FIELDNAME, new BasicDBObject( "$in", searchObjects ) );
-		return collection.find( query, projection );
+		Document query = new Document();
+		query.put( ID_FIELDNAME, new Document( "$in", searchObjects ) );
+		//return collection.find( query, projection );
```
Comment (DavideD) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: Just remove the old code, don't comment it, please

File: mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java [hunk 2]
```diff
@@ -1199,9 +1239,9 @@ private static AssociationStorageStrategy getAssociationStorageStrategy(Associat
 	public void executeBatch(OperationsQueue queue) {
 		if ( !queue.isClosed() ) {
 			Operation operation = queue.poll();
-			Map<DBCollection, BatchInsertionTask> inserts = new HashMap<DBCollection, BatchInsertionTask>();
+			Map<MongoCollection<Document>, BatchInsertionTask> inserts = new HashMap<MongoCollection<Document>, BatchInsertionTask>();
 
-			List<Tuple> insertTuples = new ArrayList<Tuple>();
+			List<Tuple> insertTuples = new LinkedList<Tuple>();
```
Comment (DavideD) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: Why did you replace an ArrayList with a LinkedList?

**Comment by DavideD** [comment]
> Also ... not close the branch please

@schernolyas The way we work is that we usually send a PR when we think it's ready to be merged, we don't send partial pull requests that don't compile or build. The exception to this rule is when the committer needs some kind of feedback. Once he has the feedback it will close the PR. There is no reason to keep a PR that cannot be merged open while you are working on it.

I cannot review your PR every time you do a single or some changes, it takes too much time.
The overall approach looks good, just reopen it or send a different one from a different branch when everything works (it compiles, builds and pass the tests). It takes 2 seconds for you to create a new branch so I don't see the problem. You can use [Travis](https://travis-ci.org/) to test your work on an external CI.

Also, it's not alright to add 3 commits to a PR and call all of them `migration to new API`, the way we describe a commit an a PR is `JIRACODE - Description`. Please, try to stick to our conventions and use a more descriptive approach. For example:

```
OGM-768 Migrate MongoDB dialect to new 3.4 API
  
  Replacing the deprecated method from the mongo driver in preparation for the GridFS support:

  * Replace DBObject with Document
  * Replace DBCollection with MongoCollection
```

Remember to delete the commented code before sending the PR, as @Sanne already mentioned before, we don't usually keep commented code around unless there is a good reason to (and you should explain the reason in the comment).

Feel free to ask questions if you have some problems with making the tests pass but don't send a PR each time, remember that if we need to check something we can always take a look at your branch.

Thanks for you help.

File: mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java (hunk 2 above)
Comment (schernolyas) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: HI @DavideD 
Because count of elements not defined. In this way LinkedList more preferable.

Comment (DavideD) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: > Because count of elements not defined. In this way LinkedList more preferable.

So this change has nothing to do with migrating the API, keep it in a separate commit, please

Comment (Sanne) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: It's a great example of why unrelated changes should be sent in a different pull request:

it's not true that a LinkedList would be more suited in this case, or at least that's my opinion so - if I were to review this PR - I would start debating about this change.
Such a debate would delay the integration of a pull request which might otherwise contain nice changes. In extreme cases, having many such "distractions" would get to endless delays, and possibly conflicts which get harder to resolve when things aren't merged quickly. It's possible that large PRs with many off-topic changes never make it to be integrated..

Comment (schernolyas) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: Dear @DavideD  and @Sanne !

Ok. I will restore it.

Comment (Sanne) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: thanks @schernolyas ! appreciate it. I hope you don't consider me being too harsh, such practices are necessary in large remote / OSS teams.. I hope you'll understand.

Personally when I had once the wrong formatting template, I simply thrown the branch away and started over.
I am not sure how large & complex your changes have been, but maybe that's easier..

Comment (schernolyas) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java]: hi @Sanne !

I will restore ArrayList. it is not a problem. 
Broken formatting .... yes You right.  Easy to restart more carefully.

About migration to 3.4 api ... It is large and complex change. And ... of course I can analyse and do self. But ... I hope to help of members because they are authors.
